### PR TITLE
Issue/5933 product images refactor take2

### PR DESCRIPTION
**Pull request by nbradbury**
This PR is a partial solution for #5933. My intent was to avoid passing a list of image models in the arguments to `ProductImagesFragment`, but that turned out to be a much larger task than I expected.

There were also other things that needed to be fixed in that fragment - for example, the `viewState` held the image list along with `productImagesState` which contained a copy of the image list, and it also held the list of uploading Uris. So this PR addresses this by moving that data out of the `viewState` and into `LiveData`.

Only one review is required, but I've tagged two developers since you both have a lot of experience with this topic.

To test:

* Go to product detail
* Tap on a product image to open `ProductImagesFragment`
* Ensure it still works as expected, including re-ordering images, uploading images, and viewing a variation's image

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt [hunk 1]
```diff
@@ -120,14 +138,30 @@ class ProductImagesViewModel @Inject constructor(
...
+            _productImages.value = _productImages.value?.let {
+                it + images
+            } ?: images
         } else {
-            viewState.copy(images = newImages)
+            _productImages.value = images
+        }
+    }
+
+    private fun addSingleImage(image: Image) {
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt:158]: NP: This is almost the same as `addImages()`. We could avoid duplicate code by replacing the body with `addImages(listOf(image))`. WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt [hunk 2]
```diff
@@ -42,17 +44,24 @@ class ProductImagesViewModel @Inject constructor(
     savedState: SavedStateHandle
 ) : ScopedViewModel(savedState) {
     private val navArgs: ProductImagesFragmentArgs by savedState.navArgs()
-    private val originalImages = navArgs.images.toList()
+    private var originalImages = navArgs.images.toList()
```
Comment (0nko) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt]: I don't think this needs to be a variable, it's not being changed anywhere.

**Review by 0nko — COMMENTED** [review]
Nice work! The images seem to be working as expected and the code looks good. I just left some minor comments in the code.

**Comment by nbradbury** [comment]
Thanks for reviewing this @0nko! I've made the changes you mentioned so ready for another round.

**Review by 0nko — APPROVED** [review]
Looking good 👍. I'll wait for @hichamboushaba for any input before merging.

**Comment by nbradbury** [comment]
@0nko I believe we need to hold off merging this. I enabled "Don't keep activities" then displayed the product images screen and backgrounded the app. When I returned to the screen all the images were lost (`viewState.productImages.value` was null).

I must be fundamentally misunderstanding how `LiveData` works, because I thought the data would be retained in this situation.

**Comment by 0nko** [comment]
@nbradbury Oh yeah, I disabled `Don't keep activities` on my emulator the other day and forgot to turn it back on.. Nice catch.

> I must be fundamentally misunderstanding how LiveData works, because I thought the data would be retained in this situation.

No, it's not, unfortunately. That's why there's the `SavedStateHandle.getLiveData()` function, which will preserve the LiveData value for you automatically.

Are you planning to eventually remove passing the actual image data in the navigation arguments? If so, you could do that first and change the logic in the ViewModel, so that the images are loaded using image IDs if they are missing.

**Comment by nbradbury** [comment]
> No, it's not, unfortunately. That's why there's the `SavedStateHandle.getLiveData()` function, which will preserve the LiveData value for you automatically.

Could adding that data to the `SavedStateHandle` lead to the same `TransactionTooLargeException` we're trying to solve here?

**Comment by 0nko** [comment]
> Could adding that data to the SavedStateHandle lead to the same TransactionTooLargeException we're trying to solve here?

Yeah. That's what the original solution did with having the data in the `ViewState`. We'll need to refactor the code so that only image IDs are passed around and the images are loaded lazily.

**Comment by nbradbury** [comment]
My apologies, @0nko. My misunderstanding of `LiveData` has caused a waste of your and my time. Lesson learned, at least. I'm simply going to close this since there's no benefit to it and it doesn't actually work.

**Pull request closed without merging**